sdks/java/extensions/sbe/src/main/java/org/apache/beam/sdk/extensions/sbe/SbeSchema.java

```diff
@@ -0,0 +1,252 @@
+ * required. See the Javadoc for the options for more details.
+ *
+ * <p>At this time, we cannot support serialization to an SBE message through IR. As a result, the
+ * {@code byte[]} output from the {@link PayloadSerializerProvider} serializer will be from a JSON
+ * representation of the message, not an SBE-serialized message. Downstream systems will need to
+ * account for this.
+ */
+@Experimental(Kind.SCHEMAS)
+public final class SbeSchema implements Serializable {
+  private static final long serialVersionUID = 1L;
+
+  private final @Nullable SerializableIr ir;
+  private final @Nullable IrOptions irOptions;
+  private final ImmutableList<SbeField> sbeFields;
+
+  private SbeSchema(
+      @Nullable SerializableIr ir,
+      @Nullable IrOptions irOptions,
+      ImmutableList<SbeField> sbeFields) {
+    this.ir = ir;
+    this.irOptions = irOptions;
+    this.sbeFields = sbeFields;
+  }
+
+  /**
+   * Creates a new {@link SbeSchema} from the given intermediate representation.
+   *
+   * <p>This makes no guarantees about the state of the returned instance. That is, it may or may
+   * not have the generated SBE schema representation, and it may or may not have translated the SBE
+   * schema into a Beam schema.
+   *
+   * @param ir the intermediate representation of the SBE schema. Modifications to the passed-in
+   *     value will not be reflected in the returned instance.
+   * @param irOptions options for configuring how to deal with cases where the desired behavior is
+   *     ambiguous.
+   * @return a new {@link SbeSchema} instance
+   */
+  public static SbeSchema fromIr(Ir ir, IrOptions irOptions) {
+    ImmutableList<SbeField> sbeFields = IrFieldGenerator.generateFields(ir, irOptions);
+
+    Ir copy =
+        new Ir(
+            ir.packageName(),
+            ir.namespaceName(),
+            ir.id(),
+            ir.version(),
+            ir.description(),
+            ir.semanticVersion(),
+            ir.byteOrder(),
+            ImmutableList.copyOf(ir.headerStructure().tokens()));
+
+    return new SbeSchema(SerializableIr.fromIr(copy), irOptions, sbeFields);
+  }
+
+  @VisibleForTesting
+  @Nullable
+  Ir getIr() {
+    return ir == null ? null : ir.ir();
+  }
+
+  @VisibleForTesting
+  @Nullable
+  IrOptions getIrOptions() {
+    return irOptions;
+  }
+
+  @VisibleForTesting
+  ImmutableList<SbeField> getSbeFields() {
+    return sbeFields;
+  }
+
+  /**
+   * Options for controlling what to do with unsigned types, specifically whether to use a higher
+   * bit count or, in the case of uint64, a string.
+   */
+  @AutoValue
+  public abstract static class UnsignedOptions implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    public abstract Boolean useMoreBitsForUint8();
+
+    public abstract Boolean useMoreBitsForUint16();
+
+    public abstract Boolean useMoreBitsForUint32();
+
+    public abstract Boolean useStringForUint64();
```

Comment:
+1 on making this configurable!

I'd prefer enumerated types over booleans here though. Booleans are restricted to two options and can't be expanded without a major refactor, but if these were enums you could easily add another option in the future (e.g. maybe map to BigDecimal or floating point). I'd suggest having an option for each bit width with the same enumerated settings like "COERCE_TO_UNSINGED", "HIGHER _BITWDITH_SIGNED", "STRING". HIGHER_BITWIDTH_SIGNED could raise for Uint64.

Reply:
Starting off with keeping the same bit count, changing to next higher bit count, converting to string, and converting to `BigDecimal`.